README.txt
$Id: README.txt,v 1.34.2.1 2010/09/28 21:45:38 alexb Exp $


"It feeds"


FEEDS
=====

An import and aggregation framework for Drupal.
http://drupal.org/project/feeds

Features
========

- Pluggable import configurations consisting of fetchers (get data) parsers
  (read and transform data) and processors (create content on Drupal).
-- HTTP upload (with optional PubSubHubbub support).
-- File upload.
-- CSV, RSS, Atom parsing.
-- Creates nodes or terms.
-- Creates lightweight database records if Data module is installed.
   http://drupal.org/project/data
-- Additional fetchers/parsers or processors can be added by an object oriented
   plugin system.
-- Granular mapping of parsed data to content elements.
- Import configurations can be piggy backed on nodes (thus using nodes to track
  subscriptions to feeds) or they can be used on a standalone form.
- Unlimited number of import configurations.
- Export import configurations to code.
- Optional libraries module support.

Requirements
============

- CTools 1.x
  http://drupal.org/project/ctools
- Job Scheduler 1.x
  http://drupal.org/project/job_scheduler
- Drupal 6.x
  http://drupal.org/project/drupal
- PHP safe mode is not supported, depending on your Feeds Importer configuration
  safe mode may cause no problems though.
- PHP 5.2.x recommended

Installation
============

- Install Feeds, Feeds Admin UI.
- To get started quick, install one or all of the following Feature modules:
  Feeds News, Feeds Import, Feeds Fast News (more info below).
- Make sure cron is correctly configured http://drupal.org/cron
- Go to import/ to import data.
- To use SimplePie parser, download SimplePie and place simplepie.inc into
  feeds/libraries. Recommended version: 1.2.
  http://simplepie.org/

Feature modules
===============

Feeds ships with three feature modules that can be enabled on
admin/build/modules or - if you are using Features - on admin/build/features.
http://drupal.org/project/features

The purpose of these modules is to provide a quick start for using Feeds. You
can either use them out of the box as they come or you can take them as samples
to learn how to build import or aggregation functionality with Feeds.

The feature modules merely contain sets of configurations using Feeds and in
some cases the modules Node, Views or Data. If the default configurations do not
fit your use case you can change them on the respective configuration pages for
Feeds, Node, Views or Data.

Here is a description of the provided feature modules:

- Feeds News -

This feature is a news aggregator. It provides a content type "Feed" that can
be used to subscribe to RSS or Atom feeds. Every item on such a feed is
aggregated as a node of the type "Feed item", also provided by the module.

What's neat about Feeds News is that it comes with a configured View that shows
a list of news items with every feed on the feed node's "View items" tab. It
also comes with an OPML importer filter that can be accessed under /import.

- Feeds Fast News -

This feature is very similar to Feeds News. The big difference is that instead
of aggregating a node for every item on a feed, it creates a database record
in a single table, thus significantly improving performance. This approach
especially starts to save resources when many items are being aggregated and
expired (= deleted) on a site.

- Feeds Import -

This feature is an example illustrating Feeds' import capabilities. It contains
a node importer and a user importer that can be accessed under /import. Both
accept CSV or TSV files as imports.

PubSubHubbub support
====================

Feeds supports the PubSubHubbub publish/subscribe protocol. Follow these steps
to set it up for your site.
http://code.google.com/p/pubsubhubbub/

- Go to admin/build/feeds and edit (override) the importer configuration you
  would like to use for PubSubHubbub.
- Choose the HTTP Fetcher if it is not already selected.
- On the HTTP Fetcher, click on 'settings' and check "Use PubSubHubbub".
- Optionally you can use a designated hub such as http://superfeedr.com/ or your
  own. If a designated hub is specified, every feed on this importer
  configuration will be subscribed to this hub, no matter what the feed itself
  specifies.

Libraries support
=================

If you are using Libraries module, you can place external libraries in the
Libraries module's search path (for instance sites/all/libraries. The only
external library used at the moment is SimplePie.

Libraries found in the libraries search path are preferred over libraries in
feeds/libraries/.

Transliteration support
=======================

If you plan to store files with Feeds - for instance when storing podcasts
or images from syndication feeds - it is recommended to enable the
Transliteration module to avoid issues with non-ASCII characters in file names.
http://drupal.org/project/transliteration

API Overview
============

See "The developer's guide to Feeds":
http://drupal.org/node/622700

Testing
=======

See "The developer's guide to Feeds":
http://drupal.org/node/622700

Debugging
=========

Set the Drupal variable 'feeds_debug' to TRUE (i. e. using drush). This will
create a file /tmp/feeds_[my_site_location].log. Use "tail -f" on the command
line to get a live view of debug output.

Note: at the moment, only PubSubHubbub related actions are logged.

Performance
===========

See "The site builder's guide to Feeds":
http://drupal.org/node/622698

Hidden settings
===============

Hidden settings are variables that you can define by adding them to the $conf
array in your settings.php file.

Name:        feeds_debug
Default:     FALSE
Description: Set to TRUE for enabling debug output to
             /DRUPALTMPDIR/feeds_[sitename].log

Name:        feeds_importer_class
Default:     'FeedsImporter'
Description: The class to use for importing feeds.

Name:        feeds_source_class
Default:     'FeedsSource'
Description: The class to use for handling feed sources.

Name:        feeds_worker_time
Default:     15
Description: Execution time for a queue worker, only effective if used with
             drupal_queue.

Name:        feeds_data_$importer_id
Default:     feeds_data_$importer_id
Description: The table used by FeedsDataProcessor to store feed items. Usually a
             FeedsDataProcessor builds a table name from a prefix (feeds_data_)
             and the importer's id ($importer_id). This default table name can
             be overridden by defining a variable with the same name.

Name:        feeds_node_batch_size
Default:     50
             The number of nodes feed node processor creates or deletes in one
             page load.

Name:        http_request_timeout
Default:     15
Description: Timeout in seconds to wait for an HTTP get request to finish.

Glossary
========

See "Feeds glossary":
http://drupal.org/node/622710


CHANGELOG.txt
// $Id: CHANGELOG.txt,v 1.176.2.13 2010/10/04 21:20:56 alexb Exp $

Feeds 7.x 2.0 XXXXXXXXXXXXXXXXXXX
---------------------------------

- Fix Feeds News tests, add a 'description' field to the  Feeds Item content
  type.
- #728534 alex_b: Remove FeedsFeedNodeProcessor. If you have used
  FeedsFeedNodeProcessor in the past, use FeedsNodeProcessor (Node Processor)
  instead now. It supports all of FeedsFeedNodeProcessor's functionality and
  more.
- #929066 alex_b: Track all imported items. Note: All views that use 'Feeds
  Item' fields or relationships need updating.
- #930018 alex_b: Don't show file upload when 'Supply path directly' is
  selected.
- #927892 alex_b: Add "Process in background" feature. Allows one-off imports to
  be processed in the background rather than using Batch API. Useful for very
  large imports.
- #929058 alex_b: Report status of source on import and delete forms, track
  last updated time on a source level.
- #928836: Set progress floating point directly. Note: fetchers and parsers
  must use $state->progress() for setting the batch progress now IF they support
  batching.
- #928728: Track source states by stage, not by plugin. Note: call signature of
  FeedsSource::state() has changed.
- Remove 6.x upgrade hooks.
- #923318: Fix Fatal error: Call to a member function import() on a non-object
  occuring on cron.
- Clean up basic settings form.
- Make getConfig() include configuration defaults.

Feeds 7.x 2.0 Alpha 1, 2010-09-29
---------------------------------

- #925842 alex_b: Support batching through directories on disk.
- #625196 mstrelan, alex_b: Fix array_merge(), array_intersect_key() warnings.
- Remove hidden setting feeds_worker_time. Use hook_cron_queue_info_alter() to
  modify this setting.
- #744660-80 alex_b: Expand batch support to fetchers and parsers.
 - Removed FeedsBatch classes in favor of FeedsResult classes.
 - Variable 'feeds_node_batch_size' is now called 'feeds_process_limit'.
 - Signature of FeedsParser::getSourceElement() changed.
 - Signature of FeedsProcessor::uniqueTargets() changed.
 - Signature of FeedsProcessor::existingItemId() changed.
 - Sigature for callbacks registered by hook_feeds_parser_sources_alter()
   changed.
 - Return value of FeedsFetcher::fetch() changed.
 - Signature and return value of FeedsParser::parse() changed.
 - Signature of FeedsProcessor::process() changed.
 - Signature of hook_feeds_after_parse() changed.
 - Signature of hook_feeds_after_import() changed.
 - Signature of hook_feeds_after_clear() changed.

Feeds 7.x 1.0 Alpha 1, 2010-09-21
---------------------------------

Equal to http://github.com/lxbarth/Feeds/commits/DRUPAL-7--1-0-alpha1

- Expire files returned by FeedsImportBatch after DRUPAL_MAXIMUM_TEMP_FILE_AGE
  seconds.
- FeedsFileFetcher: track uploaded files, delete unused files.
- yhahn: Upgrade FeedsTaxonomyProcessor.
- Remove handling of target items that are array. All target items must be
  objects now.
- Upgrade file and image mapper.
- Upgrade taxonomy mapper.
- Upgrade field mapper.
- Move plugin handling into FeedsPlugin class.
- Base level upgrade.

Feeds 6.x 1.0 Beta 6, 2010-09-16
--------------------------------

- #623432 Alex UA, dixon_, pvhee, cglusky, alex_b et al.: Mapper for emfield.
- #913672 andrewlevine: Break out CSV Parser into submethods so it is more
  easily overridable
- #853974 snoldak924, alex_b: Fix XSS vulnerabilities in module.
- #887846 ekes: Make FeedsSimplePieEnclosure (un)serialization safe.
- #908582 XiaN Vizjereij, alex_b: Fix "Cannot use object of type stdClass as
  array" error in mappers/taxonomy.inc.
- #906654 alex_b: Fix phantom subscriptions.
- #867892 alex_b: PubSubHubbub - slow down import frequency of feeds that are
  subscribed to hub.
- #908964 alex_b: Break out scheduler. Note: Features depends on Job Scheduler
  module now: http://drupal.org/project/job_scheduler
- #663860 funkmasterjones, infojunkie, alex_b et. al.: hook_feeds_after_parse().
- #755556 Monkey Master, andrewlevine, alex_b: Support saving local files in
  filefields.
- #891982 bangpound, twistor: Support Link 2.x.
- #870278 budda: Fix SQL query in taxonomy_get_term_by_name_vid().
- #795114 budda, alex_b: Taxonomy term processor doesn't require vocabulary to
  be set.

Feeds 6.x 1.0 Beta 5, 2010-09-10
--------------------------------

- #849840 adityakg, rbayliss, alex_b: Submit full mapping on every submission.
- #849834 rbayliss, alex_b: Generalize feeds_config_form() to feeds_form().
- #907064 alex_b: Track imported terms.
- #906720 alex_b: Introduce a hook_feeds_after_clear().
- #905820 tristan.oneil: Adjust delete message in FeedsDataProcessor to avoid
  misleading total numbers.
- #671538 mburak: Use CURLOPT_TIMEOUT to limit download time of feeds.
- #878002 Will White, David Goode: Support multiple sources per mapping target
  in FeedsDataProcessor.
- #904804 alex_b: Support exportable vocabularies.
- #836876 rsoden, Will White, alex_b: Add simple georss support to Common
  Syndication Parser.
- #889196 David Goode: Support for non-numeric vocabulary IDs for feature-based
  vocabularies.
- #632920 nickbits, dixon_, David Goode, alex_b et al: Inherit OG, taxonomy,
  language, user properties from parent feed node. Note: Signatures of
  FeedsProcessor::map(), existingItemId(), FeedsParser::getSourceElement()
  changed.
- #897258 TrevorBradley, alex_b: Mapping target nid.
- #873198 BWPanda, morningtime: Import multiple values to tag vocabulary.
- #872772 andrewlevine: Fix buildNode() (and node_load()) called unnecessarily.
- #873240 thsutton: Use isset() to avoid notices.
- #878528 Sutharsan: Don't show file in UI if file does not exist.
- #901798 alex_b: Fix time off in SitemapParser.
- #885724 eliotttf: Avoid array_flip() on non scalars.
- #885052 Hanno: Fix small typo in access rights.
- #863494 ekes, alex_b: Delete temporary enclosures file.
- #851194 alex_b: Featurize - move default functionality from feeds_defaults to

feeds_ui/feeds_ui.js
// $Id: feeds_ui.js,v 1.5 2010/02/23 04:59:06 alexb Exp $

Drupal.behaviors.feeds = function() {

  // Export form machine-readable JS
  $('.feed-name:not(.processed)').each(function() {
    $('.feed-name')
      .addClass('processed')
      .after(' <small class="feed-id-suffix">&nbsp;</small>');
    if ($('.feed-id').val() === $('.feed-name').val().toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/_+/g, '_') || $('.feed-id').val() === '') {
      $('.feed-id').parents('.form-item').hide();
      $('.feed-name').keyup(function() {
        var machine = $(this).val().toLowerCase().replace(/[^a-z0-9]+/g, '_').replace(/_+/g, '_');
        if (machine !== '_' && machine !== '') {
          $('.feed-id').val(machine);
          $('.feed-id-suffix').empty().append(' Machine name: ' + machine + ' [').append($('<a href="#">'+ Drupal.t('Edit') +'</a>').click(function() {
            $('.feed-id').parents('.form-item').show();
            $('.feed-id-suffix').hide();
            $('.feed-name').unbind('keyup');
            return false;
          })).append(']');
        }
        else {
          $('.feed-id').val(machine);
          $('.feed-id-suffix').text('');
        }
      });
      $('.feed-name').keyup();
    }
  });

  // Hide text in specific input fields.
  $('.hide-text-on-focus').focus(function() {
    $(this).val('');
  });


  // Hide submit buttons of .feeds-ui-hidden-submit class.
  $('input.form-submit.feeds-ui-hidden-submit').hide();

  /**
   * Tune checkboxes on mapping forms.
   * @see feeds_ui_mapping_form() in feeds_ui.admin.inc
   */

  // Attach submit behavior to elements with feeds-ui-trigger-submit class.
  $('.feeds-ui-trigger-submit').click(function() {
    // Use click, not form.submit() - submit() would use the wrong submission
    // handler.
    $('input.form-submit.feeds-ui-hidden-submit').click();
  });

  // Replace checkbox with .feeds-ui-checkbox-link class with a link.
  $('.feeds-ui-checkbox-link:not(.processed)').each(function(i) {
    $(this).addClass('processed').after(
      '<a href="#" onclick="return false;" class="feeds-ui-trigger-remove">' + $('label', this).text() + '</a>'
    ).hide();
  });

  // Check the box and then submit.
  $('.feeds-ui-trigger-remove').click(function() {
    // Use click, not form.submit() - submit() would use the wrong submission
    // handler.
    $(this).prev().children().children().children().attr('checked', 1);
    $('input.form-submit.feeds-ui-hidden-submit').click();
  });

  // Replace radio with .feeds-ui-radio-link class with a link.
  $('.feeds-ui-radio-link:not(.processed)').parent().each(function(i) {
    checked = '';
    if ($(this).children('input').attr('checked')) {
      checked = ' checked';
    }
    $(this).addClass('processed').after(
      '<a href="#" onclick="return false;" class="feeds-ui-check-submit' + checked + '" id="' + $(this).children('input').attr('id') + '">' + $(this).parent().text() + '</a>'
    );
    $(this).hide();
  });

  // Hide the the radio that is selected.
  $('.feeds-ui-check-submit.checked').parent().hide();

  // Check the radio and then submit.
  $('.feeds-ui-check-submit').click(function() {
    // Use click, not form.submit() - submit() would use the wrong submission
    // handler.
    $('#' + $(this).attr('id')).attr('checked', 1);
    $('input.form-submit.feeds-ui-hidden-submit').click();
  });

  // Show item link settings conditionally.
  // @todo Generalize dependencies between form elements.
  // @todo Refuse submission if required form elements are missing.
  var path_label = $('#edit-item-link-path-wrapper').children('label').text();
  var name_label = $('#edit-item-link-name-wrapper').children('label').text();
  if ($('#edit-item-link').attr('checked')) {
    $('#edit-item-link-path').show();
    $('#edit-item-link-path-wrapper').children('label').append('<span class="form-required">*</span>');
    $('#edit-item-link-name').show();
    $('#edit-item-link-name-wrapper').children('label').append('<span class="form-required">*</span>');
  }
  else {
    $('#edit-item-link-path').hide();
    $('#edit-item-link-name').hide();
  }
  $('#edit-item-link').click(function() {
    if ($('#edit-item-link').attr('checked')) {
      $('#edit-item-link-path-wrapper').show(100);
      $('#edit-item-link-path-wrapper').children('label').append('<span class="form-required">*</span>');
      $('#edit-item-link-name-wrapper').show(100);
      $('#edit-item-link-name-wrapper').children('label').append('<span class="form-required">*</span>');
    }
    else {
      $('#edit-item-link-path-wrapper').hide(100);
      $('#edit-item-link-path-wrapper').children('label').text(path_label);
      $('#edit-item-link-name-wrapper').hide(100);
      $('#edit-item-link-name-wrapper').children('label').text(name_label);
    }
  });

  // Show pubsub settings conditionally.
  if ($('#edit-use-pubsubhubbub').attr('checked')) {
    $('#edit-designated-hub-wrapper').show();
  }
  else {
    $('#edit-designated-hub-wrapper').hide();
  }
  $('#edit-use-pubsubhubbub').click(function() {
    if ($(this).attr('checked')) {
      $('#edit-designated-hub-wrapper').show(100);
    }
    else {
      $('#edit-designated-hub-wrapper').hide(100);
    }
  });
};


feeds.api.php
<?php
// $Id: feeds.api.php,v 1.17.2.1 2010/09/29 23:56:09 alexb Exp $

/**
 * @mainpage
 *
 * @verbinclude README.txt
 */

/**
 * Feeds offers a CTools based plugin API. Fetchers, parsers and processors are
 * declared to Feeds as plugins.
 *
 * @see feeds_feeds_plugins()
 * @see FeedsFetcher
 * @see FeedsParser
 * @see FeedsProcessor
 *
 * @defgroup pluginapi Plugin API
 * @{
 */

/**
 * Example of a CTools plugin hook that needs to be implemented to make
 * hook_feeds_plugins() discoverable by CTools and Feeds. The hook specifies
 * that the hook_feeds_plugins() returns Feeds Plugin API version 1 style
 * plugins.
 */
function hook_ctools_plugin_api($owner, $api) {
  if ($owner == 'feeds' && $api == 'plugins') {
    return array('version' => 1);
  }
}

/**
 * A hook_feeds_plugins() declares available Fetcher, Parser or Processor
 * plugins to Feeds. For an example look at feeds_feeds_plugin(). For exposing
 * this hook hook_ctools_plugin_api() MUST be implemented, too.
 *
 * @see feeds_feeds_plugin()
 */
function hook_feeds_plugins() {
  $info = array();
  $info['MyFetcher'] = array(
    'name' => 'My Fetcher',
    'description' => 'Fetches my stuff.',
    'help' => 'More verbose description here. Will be displayed on fetcher selection menu.',
    'handler' => array(
      'parent' => 'FeedsFetcher',
      'class' => 'MyFetcher',
      'file' => 'MyFetcher.inc',
      'path' => drupal_get_path('module', 'my_module'), // Feeds will look for MyFetcher.inc in the my_module directory.
    ),
  );
  $info['MyParser'] = array(
    'name' => 'ODK parser',
    'description' => 'Parse my stuff.',
    'help' => 'More verbose description here. Will be displayed on parser selection menu.',
    'handler' => array(
      'parent' => 'FeedsParser', // Being directly or indirectly an extension of FeedsParser makes a plugin a parser plugin.
      'class' => 'MyParser',
      'file' => 'MyParser.inc',
      'path' => drupal_get_path('module', 'my_module'),
    ),
  );
  $info['MyProcessor'] = array(
    'name' => 'ODK parser',
    'description' => 'Process my stuff.',
    'help' => 'More verbose description here. Will be displayed on processor selection menu.',
    'handler' => array(
      'parent' => 'FeedsProcessor',
      'class' => 'MyProcessor',
      'file' => 'MyProcessor.inc',
      'path' => drupal_get_path('module', 'my_module'),
    ),
  );
  return $info;
}

/**
 * @}
 */

/**
 * @defgroup import Import and clear hooks
 * @{
 */

/**
 * Invoked after a feed source has been parsed, before it will be processed.
 *
 * @param $source
 *  FeedsSource object that describes the source that has been imported.
 * @param $result
 *   FeedsParserResult object that has been parsed from the source.
 */
function hook_feeds_after_parse(FeedsSource $source, FeedsParserResult $result) {
  // For example, set title of imported content:
  $result->setTitle('Import number ' . my_module_import_id());
}

/**
 * Invoked after a feed source has been imported.
 *
 * @param $source
 *  FeedsSource object that describes the source that has been imported.
 */
function hook_feeds_after_import(FeedsSource $source) {
  // See geotaxonomy module's implementation for an example.
}

/**
 * Invoked after a feed source has been cleared of its items.
 *
 * @param $source
 *  FeedsSource object that describes the source that has been cleared.
 */
function hook_feeds_after_clear(FeedsSource $source) {
}

/**
 * @}
 */

/**
 * @defgroup mappingapi Mapping API
 * @{
 */

/**
 * Alter mapping sources.
 *
 * Use this hook to add additional mapping sources for any parser. Allows for
 * registering a callback to be invoked at mapping time.
 *
 * @see my_source_get_source().
 * @see locale_feeds_parser_sources_alter().
 */
function hook_feeds_parser_sources_alter(&$sources, $content_type) {
  $sources['my_source'] = array(
    'name' => t('Images in description element'),
    'description' => t('Images occuring in the description element of a feed item.'),
    'callback' => 'my_source_get_source',
  );
}

/**
 * Example callback specified in hook_feeds_parser_sources_alter().
 *
 * To be invoked on mapping time.
 *
 * @param $source
 *   The FeedsSource object being imported.
 * @param $result
 *   The FeedsParserResult object being mapped from.
 * @param $key
 *   The key specified in the $sources array in
 *   hook_feeds_parser_sources_alter().
 *
 * @return
 *   The value to be extracted from the source.
 *
 * @see hook_feeds_parser_sources_alter().
 * @see locale_feeds_get_source().
 */
function my_source_get_source(FeedsSource $source, FeedsParserResult $result, $key) {
  $item = $result->currentItem();
  return my_source_parse_images($item['description']);
}

/**
 * Alter mapping targets for entities. Use this hook to add additional target
 * options to the mapping form of Node processors.
 *
 * If the key in $targets[] does not correspond to the actual key on the node
 * object ($node->key), real_target MUST be specified. See mappers/link.inc
 *
 * For an example implementation, see mappers/content.inc
 *
 * @param &$targets
 *   Array containing the targets to be offered to the user. Add to this array
 *   to expose additional options. Remove from this array to suppress options.
 *   Remove with caution.
 * @param $entity_type
 *   The entity type of the target, for instance a 'node' entity.
 * @param $bundle_name
 *   The bundle name for which to alter targets.
 */
function hook_feeds_processor_targets_alter(&$targets, $entity_type, $bundle_name) {
  if ($entity_type == 'node') {
    $targets['my_node_field'] = array(
      'name' => t('My custom node field'),
      'description' => t('Description of what my custom node field does.'),
      'callback' => 'my_module_set_target',
    );
    $targets['my_node_field2'] = array(
      'name' => t('My Second custom node field'),
      'description' => t('Description of what my second custom node field does.'),
      'callback' => 'my_module_set_target2',
      'real_target' => 'my_node_field_two', // Specify real target field on node.
    );
  }
}

/**
 * Example callback specified in hook_feeds_processor_targets_alter().
 *
 * @param $entity
 *   An entity object, for instance a node object.
 * @param $target
 *   A string identifying the target on the node.
 * @param $value
 *   The value to populate the target with.
 *
 */
function my_module_set_target($entity, $target, $value) {
  $entity->$target['und'][0]['value'] = $value;
}

/**
 * @}
 */
